src/aspire/nfft/cufinufft.py

```diff
@@ -0,0 +1,131 @@
+from ctypes import c_int
+
+import logging
+import numpy as np
+import pycuda.autoinit
+import pycuda.driver as cuda
+import pycuda.gpuarray as gpuarray
+from cufinufft import cufinufft
+
+from aspire.nfft import Plan
+from aspire.utils import ensure
+
+logger = logging.getLogger(__name__)
+
+class cuFINufftPlan(Plan):
+    def __init__(self, sz, fourier_pts, epsilon=1e-15, many=0, **kwargs):
+        """
+        A plan for non-uniform FFT in 2D or 3D.
+
+        :param sz: A tuple indicating the geometry of the signal
+        :param fourier_pts: The points in Fourier space where the Fourier transform is to be calculated,
+            arranged as a dimension-by-K array. These need to be in the range [-pi, pi] in each dimension.
+        :param epsilon: The desired precision of the NUFFT
+        :param many: Optional integer indicating if you would like to compute a batch of `many`
+        transforms.  Implies vol_f.shape is (..., `many`). Defaults to 0 which disables batching.
+        """
+
+        # Passing "many" expects one large higher dimensional array.
+        #   Set some housekeeping variables so we can discern how to handle the dims later.
+        self.many = False
+        self.ntransforms = 1
+        if many != 0:
+            self.many = True
+            self.ntransforms = many
+
+        # Basic dtype passthough.
+        dtype = fourier_pts.dtype
+        if dtype == np.float64 or dtype == np.complex128:
+            self.dtype = np.float64
+            self.complex_dtype = np.complex128
+        elif dtype == np.float32 or dtype == np.complex64:
+            self.dtype = np.float32
+            self.complex_dtype = np.complex64
+        else:
+            raise RuntimeError("Unsupported dtype encountered")
+
+        self.sz = sz
+        self.dim = len(sz)
+        # TODO: Following the row major conversion, this should no longer happen.
+        if fourier_pts.flags.f_contiguous:
+            logger.warning('cufinufft has caught an F_CONTIGUOUS array,'
+                           ' `fourier_pts` will be copied to C_CONTIGUOUS.')
+        self.fourier_pts = np.asarray(np.mod(fourier_pts + np.pi, 2 * np.pi) - np.pi, order='C', dtype=self.dtype)
+        self.num_pts = fourier_pts.shape[1]
+        self.epsilon = max(epsilon, np.finfo(self.dtype).eps)
+
+        self._transform_plan = cufinufft(2, self.sz, -1, self.epsilon, ntransforms=self.ntransforms,
+                                         dtype=self.dtype)
+
+        self._adjoint_plan = cufinufft(1, self.sz, 1, self.epsilon, ntransforms=self.ntransforms,
+                                       dtype=self.dtype)
```

Comment:
Here we do not yet know if user wants to set points for transform, adjoint, or both.  I can set points for both here if that is somehow better then setting them per method later. You will still have two calls, because there are two plans the way this was wrapped up.

Reply:
I cleaned it up a little. I agree it looks cleaner now.

Reply:
Right. We're suffering a little bit here since we're prepping both forward and adjoint, but I don't think it's a huge deal in the grand scheme of things (IIRC plan instantiation and setting points are comparable in running time). If it turns out to be an issue, we can look into forward- or adjoint-specific plans in the future.